### PR TITLE
Relax day ahead persistence midnight to midnight (UTC) requirement

### LINKED ISSUE #471: Use site timezone when checking midnight to midnight for day ahead persistence
Sentry Issue: [WORKER-API-E](https://sentry.io/organizations/solararbiter-rx/issues/1705403200/?referrer=github_integration)

```
ValueError: Day ahead persistence requires midnight to midnight periods
(2 additional frame(s) were not displayed)
...
  File "rq/job.py", line 651, in _execute
    return self.func(*self.args, **self.kwargs)
  File "sfa_api/jobs.py", line 160, in execute_job
    token, max_run_time, base_url=base_url)
  File "solarforecastarbiter/reference_forecasts/main.py", line 606, in make_latest_persistence_forecasts
    index=index)
  File "solarforecastarbiter/reference_forecasts/main.py", line 239, in run_persistence
    utils._check_midnight_to_midnight(forecast_start, forecast_end)
  File "solarforecastarbiter/reference_forecasts/utils.py", line 146, in _check_midnight_to_midnight
    'Day ahead persistence requires midnight to midnight periods')
```

### PR DESCRIPTION
**Pull request by alorenzo175**
- [x] Closes #471.
  - [x] I am familiar with the [contributing guidelines](https://solarforecastarbiter-core.readthedocs.io/en/latest/contributing.html).
  - [x] Tests added.
  - [x] Updates entries to [`docs/source/api.rst`](https://github.com/SolarArbiter/solarforecastarbiter-core/blob/master/docs/source/api.rst) for API changes.
  - [x] Adds descriptions to appropriate "what's new" file in [`docs/source/whatsnew`](https://github.com/SolarArbiter/solarforecastarbiter-core/tree/master/docs/source/whatsnew) for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` or this Pull Request with `` :pull:`num` ``. Includes contributor name and/or GitHub username (link with `` :ghuser:`user` ``).
  - [x] New code is fully documented. Includes [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) compliant docstrings, examples, and comments where necessary.
  - [x] Maintainer: Appropriate GitHub Labels and Milestone are assigned to the Pull Request and linked Issue.

also addresses an issue of how issue times were issued that changed based on the timezone. Forecast issues times are now explicitly expected to be in UTC. Also fixes and adds tests to make sure that weekahead  persistence always uses data from 1 week before the forecast valid times.

### CODE REVIEW

**Comment by alorenzo175** [comment]
When reviewing, I think we should make sure the test behaviour is exactly what we expect.

File: solarforecastarbiter/reference_forecasts/utils.py [hunk 1]
```diff
@@ -191,30 +192,32 @@ def _dayahead_start_end(run_time):
     until end of day. So, forecast *never* uses obs > 24 hr old at each
     valid time. Arguably too much for a reference forecast.
     """
-
-    data_end = run_time.floor('1d')
+    # data_end = last forecast time for next issue of run - 1 day
+    data_end = issue_time + forecast.lead_time_to_start + forecast.run_length
+    while data_end > run_time:
+        data_end -= pd.Timedelta('1d')
```
Comment (wholmgren) [inline solarforecastarbiter/reference_forecasts/utils.py:198]: Is this meant to ensure that 1. `data_end` is always less than `run_time` and 2. `data_end.time()` doesn't change?

Can the same thing can be accomplished more efficiently with something like

```python
if data_end > run_time:
    day_offset = (data_end - run_time).days
    data_end -= pd.Timedelta(days=day_offset)
```

Comment (wholmgren) [reply solarforecastarbiter/reference_forecasts/utils.py:198]: maybe I'm misunderstanding something since I would have thought the same requirement would be added to `_weekahead_start_end`.

File: solarforecastarbiter/reference_forecasts/main.py [hunk 2]
```diff
@@ -253,13 +255,18 @@ def run_persistence(session, observation, forecast, run_time, issue_time,
         forecast, issue_time, False)
     intraday = utils._is_intraday(forecast)
     if not intraday:
-        # raise ValueError if not intraday and not midnight to midnight
-        utils._check_midnight_to_midnight(forecast_start, forecast_end)
+        if forecast_end - forecast_start > pd.Timedelta('1d'):
+            raise ValueError(
+                'Persistence forecasts with longer than 1 day are '
+                'not currently supported')
```
[the remark this answers is not in the document]
Comment (alorenzo175) [reply solarforecastarbiter/reference_forecasts/main.py:261]: I thought there was a restriction somewhere else like `utils.check_persistence_compatibility` and it was part of midnight to midnight. I don't see it anywhere else, so can remove

File: solarforecastarbiter/reference_forecasts/utils.py (hunk 1 above)
Comment (alorenzo175) [reply solarforecastarbiter/reference_forecasts/utils.py]: Actually, I think what's important is that `data_end < issue_time` and `data[i].time() == forecast[i].time()` (so agreed on 2). I've updated to enforce this, although it can lead to quirks like a 23h ahead forecast using more recent data than a 1h ahead. 

As for weekahead, there it's important that the day of the week along with time is maintained. So just subtracting the week from the forecast start and assuming run_time is after data end should be ok. If a set of parameters always violates that, the parameters should be changed.

File: solarforecastarbiter/reference_forecasts/tests/test_fx_utils.py [hunk 3]
```diff
@@ -256,40 +281,104 @@ def test_get_data_start_end_labels_obs_avg_fx_instant(site_metadata):
...
 
 
-@pytest.mark.parametrize("variable,expected_start,expected_end", [
+@pytest.mark.parametrize('rl,expected_start,expected_end,rt,lt', [
+    ('1d', '20190409T0000Z', '20190410T0000Z', '20190410T2230Z', '1h'),
+    ('1d', '20190409T0000Z', '20190410T0000Z', '20190410T0001Z', '1h'),
+    ('1d', '20190409T0000Z', '20190410T0000Z', '20190410T0001Z', '25h'),
+    ('1d', '20190409T0000Z', '20190410T0000Z', '20190410T0001Z', '49h'),
+    ('1d', '20190408T2300Z', '20190409T2300Z', '20190410T2230Z', '24h'),
+    # quirky that 23h ahead uses more recent data than 1h ahead, but
```
Comment (wholmgren) [inline solarforecastarbiter/reference_forecasts/tests/test_fx_utils.py:294]: It is quirky but I think I've convinced myself that it is correct. My understanding is:

`rt = '20190410T2230Z'`, `issue_time = '20190410T2300Z'`,  `lead_time_to_start = '23h'`, so fx from `20190411T2200Z` to `20190412T2200Z`. The most recent data I can use for that 24 hr period is `20190409T2200Z` to `20190410T2200Z`.

But if I set `rt = '20190410T2130Z'`, the most recent data I can use for that 24 hr period is `20190408T2200Z` to `20190409T2200Z`.

File: solarforecastarbiter/reference_forecasts/tests/test_fx_utils.py [hunk 4]
```diff
@@ -256,40 +281,104 @@ def test_get_data_start_end_labels_obs_avg_fx_instant(site_metadata):
...
+    # timestamps are maintained and end is before issue time
+    ('1d', '20190409T2200Z', '20190410T2200Z', '20190410T2230Z', '23h'),
+    ('2d', '20190408T0000Z', '20190410T0000Z', '20190410T2230Z', '1h'),
+    ('36h', '20190409T0000Z', '20190410T1200Z', '20190410T2300Z', '1h'),
+    # run time is too early compared to init time so data_end > run_time
```
Comment (wholmgren) [inline solarforecastarbiter/reference_forecasts/tests/test_fx_utils.py:299]: issue time, not init time

File: solarforecastarbiter/reference_forecasts/tests/test_fx_utils.py [hunk 5]
```diff
@@ -256,40 +281,104 @@ def test_get_data_start_end_labels_obs_avg_fx_instant(site_metadata):
...
+    # and that's ok.
```
Comment (wholmgren) [inline solarforecastarbiter/reference_forecasts/tests/test_fx_utils.py:300]: I don't understand why that's ok. There would not be enough data to issue a complete forecast.

File: solarforecastarbiter/reference_forecasts/tests/test_fx_utils.py (hunk 3 above)
Comment (wholmgren) [reply solarforecastarbiter/reference_forecasts/tests/test_fx_utils.py:294]: or maybe I don't understand (see comment on "and that's ok" below)

File: solarforecastarbiter/reference_forecasts/main.py [hunk 6]
```diff
@@ -252,14 +251,14 @@ def run_persistence(session, observation, forecast, run_time, issue_time,
...
-        utils._check_midnight_to_midnight(forecast_start, forecast_end)
 
     if load_data is None:
         load_data = _default_load_data(session)
     data_start, data_end = utils.get_data_start_end(
-        observation, forecast, run_time)
+        observation, forecast, run_time, issue_time)
+    if data_end > run_time:
+        raise ValueError(
+            'Persistence forecast requires data from after run_time')
```
[the remark this answers is not in the document]
Comment (alorenzo175) [reply solarforecastarbiter/reference_forecasts/main.py:261]: They don't at the moment and I'm not sure they should. I should add documentation for this but I think at least day ahead and week ahead should not rely on run time to set data start/end. Basically, you are able to determine what data went into the persistence forecast without worrying about when it was made. Therefore, it's possible that the run time is not after data end, so fail instead of not having enough data for the forecast.

File: solarforecastarbiter/reference_forecasts/utils.py [hunk 7]
```diff
@@ -28,9 +28,13 @@ def get_issue_times(forecast, start_from=None):
         issue = pd.Timestamp.combine(pd.Timestamp(0).date(),
                                      forecast.issue_time_of_day)
     else:
-        issue = pd.Timestamp.combine(start_from.date(),
+        if start_from.tz is not None:
+            start_date = start_from.astimezone('UTC').date()
+        else:
+            start_date = start_from.date()
+        issue = pd.Timestamp.combine(start_date,
```
[the remark this answers is not in the document]
Comment (alorenzo175) [reply solarforecastarbiter/reference_forecasts/utils.py:35]: probably. should check and test that

File: solarforecastarbiter/reference_forecasts/tests/test_fx_utils.py [hunk 8]
```diff
@@ -36,7 +36,20 @@ def test_issue_times(single_forecast, issuetime, rl, lt, expected):
       pd.Timestamp('20190102T0500Z')]),
     ('11:00', '12h', '3h', pd.Timestamp('20190101T2300Z'),
      [pd.Timestamp('20190101T1100Z'), pd.Timestamp('20190101T2300Z'),
-      pd.Timestamp('20190102T1100Z')])
+      pd.Timestamp('20190102T1100Z')]),
+    ('05:00', '12h', '1h', pd.Timestamp('20190101T0000Z'),
+     [pd.Timestamp('20190101T0500Z'), pd.Timestamp('20190101T1700Z'),
+      pd.Timestamp('20190102T0500Z')]),
+    ('05:00', '12h', '1h', pd.Timestamp('20190101T0900-06:00'),
+     [pd.Timestamp('20181231T2300-06:00'), pd.Timestamp('20190101T1100-06:00'),
```
[the remark this answers is not in the document]
Comment (alorenzo175) [reply solarforecastarbiter/reference_forecasts/tests/test_fx_utils.py:44]: yes, that is the purpose. I'll update the documentation

File: solarforecastarbiter/reference_forecasts/tests/test_fx_utils.py [hunk 9]
```diff
@@ -85,8 +99,9 @@ def test_get_data_start_end_labels_1h_window_limit(site_metadata):
         interval_label='beginning')
```
[the remark this answers is not in the document]
Comment (alorenzo175) [reply solarforecastarbiter/reference_forecasts/tests/test_fx_utils.py:99]: I find it easier to have conftest.py open next to this file rather than trying to scroll all the time to check the defaults

File: solarforecastarbiter/reference_forecasts/tests/test_fx_utils.py [hunk 10]
```diff
@@ -177,14 +197,15 @@ def test_get_data_start_end_labels_obs_fx_instant(site_metadata):
         interval_length=pd.Timedelta('5min'), interval_label='instant')
     forecast = default_forecast(
         site_metadata,
-        issue_time_of_day=dt.time(hour=5),
+        issue_time_of_day=dt.time(hour=23),
```
[the remark this answers is not in the document]
Comment (alorenzo175) [reply solarforecastarbiter/reference_forecasts/tests/test_fx_utils.py:200]: yes, confirmed

File: solarforecastarbiter/reference_forecasts/tests/test_fx_utils.py [hunk 11]
```diff
@@ -256,40 +281,104 @@ def test_get_data_start_end_labels_obs_avg_fx_instant(site_metadata):
...
+def test_get_data_start_end_time_dayahead(site_metadata, rl, rt, lt,
+                                          expected_start, expected_end):
+    observation = default_observation(
+        site_metadata,
+        interval_length=pd.Timedelta('5min'), interval_label='beginning'
+    )
+
+    run_time = pd.Timestamp(rt)
+    issue_time = pd.Timestamp('20190410T2300Z')
+    # fx from 2019-04-11 00:00
```
[the remark this answers is not in the document]
Comment (alorenzo175) [reply solarforecastarbiter/reference_forecasts/tests/test_fx_utils.py:313]: should have been removed. It does vary with lt but definitely should not with rt, i.e. run time should not change data start/end for a given issue time

File: solarforecastarbiter/reference_forecasts/tests/test_fx_utils.py (hunk 5 above)
Comment (alorenzo175) [reply solarforecastarbiter/reference_forecasts/tests/test_fx_utils.py:300]: It's ok because run persistence raises that value error. Day/week ahead persistence data is set based on the forecast parameters, so this is basically just testing that you tried to make the forecast too early

File: solarforecastarbiter/reference_forecasts/tests/test_fx_utils.py (hunk 3 above)
Comment (alorenzo175) [reply solarforecastarbiter/reference_forecasts/tests/test_fx_utils.py:294]: Again, more clear if I stated that the goal is to set the data period based on the forecast parameters and not based on when one tried to make a forecast. So the data period is always `20190409T2200Z` to `20190410T2200Z` for the `issue_time = '20190410T2300Z'` without regard for the run time. `run_persistence` has the safeguard so that it only produces the forecast for that issue time when `run_time > 20190410T2200Z`

File: solarforecastarbiter/reference_forecasts/utils.py [hunk 12]
```diff
@@ -28,9 +28,13 @@ def get_issue_times(forecast, start_from=None):
         issue = pd.Timestamp.combine(pd.Timestamp(0).date(),
                                      forecast.issue_time_of_day)
     else:
```
[the remark this answers is not in the document]
Comment (alorenzo175) [reply solarforecastarbiter/reference_forecasts/utils.py:30]: hmm, I think anything calling it without start_time has been refactored or otherwise removed, so this can require it

File: solarforecastarbiter/reference_forecasts/main.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (wholmgren) [reply solarforecastarbiter/reference_forecasts/main.py:261]: that makes sense and now I agree

**Comment by wholmgren** [comment]
Thanks for the responses and the improved documentation. It makes a lot more sense now. Minor comments on `test_main.py`... 

Do you feel confident about the observation and forecast `interval_label` behavior with these functions? Most of the tests use matched instantaneous (required) or matched beginning. Not a lot of ending or beginning/ending mismatch. I think it should be ok because all of the code you're modifying uses timestamps that represent instants instead of intervals.

**Comment by alorenzo175** [comment]
> Thanks for the responses and the improved documentation. It makes a lot more sense now. Minor comments on `test_main.py`...
> 
> Do you feel confident about the observation and forecast `interval_label` behavior with these functions? Most of the tests use matched instantaneous (required) or matched beginning. Not a lot of ending or beginning/ending mismatch. I think it should be ok because all of the code you're modifying uses timestamps that represent instants instead of intervals.

I am confident, but added some tests anyway

**Comment by alorenzo175** [comment]
OK, I rewrote `get_issue_times` to hopefully be easier to understand and also deal with run lenghts > 1d. It should now return all issue times for the day of start_from and the first one from the next day.

File: solarforecastarbiter/reference_forecasts/tests/test_main.py [hunk 13]
```diff
@@ -155,13 +155,101 @@ def test_run_persistence_interval(session, site_metadata, obs_5min_begin,
...
+    assert out.loc['20190105T0000Z'] == 0
+    assert out.loc['20190105T0100Z'] == 1
+    assert out.loc['20190105T0200Z'] == 2
+    assert main.persistence.persistence_interval.call_count == 1
+
+
+def test_run_persistence_interval_tz(session, site_metadata, obs_5min_begin,
+                                     mocker):
+    run_time = pd.Timestamp('20190103T0345Z')
+    site = site_metadata.replace(timezone='Etc/GMT+5')
```
[the remark this answers is not in the document]
Comment (alorenzo175) [reply solarforecastarbiter/reference_forecasts/tests/test_main.py:221]: I guess it was more about not midnight, but now it does test localized times.

File: solarforecastarbiter/reference_forecasts/utils.py [hunk 14]
```diff
@@ -1,45 +1,50 @@
...
+    earliest_start = pd.Timestamp.combine(
+        (start_from - dayadj).date(), start_time)
+    possible_times = []
+    for i in range(3):
+        start = earliest_start + i * dayadj
+        end = (start + dayadj).floor('1d')
+        possible_times.extend(list(
+            pd.date_range(start=start, end=end, freq=forecast.run_length)))
+    # broad range of times that should cover start_from and next time
+    # even after timezone conversion
```
Comment (wholmgren) [inline solarforecastarbiter/reference_forecasts/utils.py:41]: the lines above made more sense once I got to this comment. I don't understand the motivation for this approach though. It's also not clear that this works if `forecast.run_length` is very short.

Comment (alorenzo175) [reply solarforecastarbiter/reference_forecasts/utils.py:41]: for example if issue time is 06:00 UTC and start = 20200101T0000Z with a run_length of 1 hour, 2020-01-01T00:00Z is a valid start time that should be returned. As it's written we need to include the previous day (line 33) to generate that issue time. We also need to get the issue times for the next day to get the first one (2020-01-02T06:00Z), hence  i in range(3) (yesterday, today, tomorrow). It is certainly too broad, but all the edge cases are handled. 

The tests show that it does work for short run lengths (many are 1hour, shorter would be fine too) because of the `ceil` on line 31.
